### DISCUSSION TITLE
Show only top-level models in the docs

### DISCUSSION THREAD

**Discussion by KDruzhkin** [fastapi/fastapi#11859] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel
from fastapi import FastAPI


class _X(BaseModel):
    x: int


class _Y(BaseModel):
    y: int
    x_model: _X


class Z(BaseModel):
    z: int
    y_model: _Y


app = FastAPI()


@app.get("/")
def read_root(z_model: Z) -> int:
    return 0
```

### Description

I am looking at the auto-generated docs page <http://127.0.0.1:8000/docs>, section “Schemas”. It shows not only the main model **Z**, but also auxiliary models **_Y** and **_X**. How do I hide them?

Note that the _schemas_ of **_Y** and **_X** are perfectly visible inside the definition of **Z**. I only want to hide their _names_, and to shorten the list of schemas.

### Operating System

macOS

### FastAPI Version

0.111.1

### Pydantic Version

2.8.2

### Python Version

3.12

### Additional Context

![docs](https://github.com/user-attachments/assets/3f95bbea-9169-4255-bf46-8c2fe098a99b)

**Comment by AIdjis** [comment]
Hello, if you want to hide models from the schema documentation, you might need to manually create a custom OpenAPI schema to hide them, just like the example below:


```python 
from pydantic import BaseModel
from fastapi import FastAPI
from fastapi.openapi.utils import get_openapi



class _X(BaseModel):
    x: int


class _Y(BaseModel):
    y: int
    x_model: _X


class Z(BaseModel):
    z: int
    y_model: _Y


app = FastAPI()


@app.get("/")
def read_root(z_model: Z) -> int:
    return 0

def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
        title="My API",
        version="1",
        routes=app.routes,
    )
    # Remove auxiliary models from the schema
    openapi_schema["components"]["schemas"].pop("_X", None)
    openapi_schema["components"]["schemas"].pop("_Y", None)
    app.openapi_schema = openapi_schema
    return app.openapi_schema
app.openapi = custom_openapi

```
You might encounter an error in the documentation related to the OpenAPI schema. It still references the _Y and _X models in the Z model, even though they have been removed from the schemas. just ignore it.

Remember, this technique only removes the models from the documentation. You still need to ensure that the data is properly sent to your API for it to function correctly.

**Reply by KDruzhkin** [reply]
It works, but the error is shown above, as you expected.
![error](https://github.com/user-attachments/assets/630860f2-a1f1-4787-9f2f-848646c9abb9)

**Reply by AIdjis** [reply]
I mentioned that you can ignore this error. I've tested the endpoint, and it works correctly. You just need to ensure that you send the JSON data properly. This is the perfect solution to your question. FastAPI requires that schemas adhere strictly to the schema definitions provided in the code. To address this, you need to set up custom schema documentation.

**Comment by YuriiMotov** [comment, marked as answer]
What you need is to embed the schema of `_X` into `_Y` and then embed schema of `_Y` into `Z`.

You can achieve this with the code like below (it hasn't been tested well, it's just to give you an idea):

```python
import copy
from pydantic import BaseModel
from fastapi import FastAPI
from fastapi.openapi.utils import get_openapi


class _X(BaseModel):
    x: int


class _Y(BaseModel):
    y: int
    x_model: _X


class Z(BaseModel):
    z: int
    y_model: _Y


app = FastAPI()


@app.get("/")
def read_root(z_model: Z) -> int:
    return 0


def extract_used_refs(node: dict | list, used: set):
    """Recursively find all $ref targets in a schema node."""
    if isinstance(node, dict):
        if "$ref" in node:
            ref = node["$ref"]
            if ref.startswith("#/components/schemas/"):
                model_name = ref.split("/")[-1]
                used.add(model_name)
        for value in node.values():
            extract_used_refs(value, used)
    elif isinstance(node, list):
        for item in node:
            extract_used_refs(item, used)

def inline_refs_rec(node: dict | list, components: dict) -> dict | list:
    """Recursively replace $ref with the referenced schema."""
    if isinstance(node, dict):
        if "$ref" in node and node["$ref"].startswith("#/components/schemas/"):
            ref_name = node["$ref"].split("/")[-1]
            resolved = inline_refs_rec(copy.deepcopy(components[ref_name]), components)
            # Merge any other keys (like title overrides)
            return {**resolved, **{k: v for k, v in node.items() if k != "$ref"}}
        return {k: inline_refs_rec(v, components) for k, v in node.items()}
    elif isinstance(node, list):
        return [inline_refs_rec(item, components) for item in node]
    return node

def inline_refs(schema: dict) -> dict:
    schema_copy = copy.deepcopy(schema)
    components = schema_copy.get("components", {}).get("schemas", {})
    if not components:
        return schema_copy  # No schemas to process

    # Step 1: inline all $ref's across components
    if "components" in schema_copy and "schemas" in schema_copy["components"]:
        for key in list(schema_copy["components"]["schemas"].keys()):
            schema_copy["components"]["schemas"][key] = inline_refs_rec(
                schema_copy["components"]["schemas"][key], components
            )

    # Step 2: collect all used $ref models
    used_models = set()
    extract_used_refs(schema_copy.get("paths", {}), used_models)
    for schema_def in components.values():
        extract_used_refs(schema_def, used_models)

    # Step 3: keep only used models
    schema_copy["components"]["schemas"] = {
        name: components[name] for name in used_models if name in components
    }

    return schema_copy


def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
        title=app.title,
        version=app.version,
        routes=app.routes,
    )

    openapi_schema = inline_refs(openapi_schema)

    app.openapi_schema = openapi_schema
    return app.openapi_schema

app.openapi = custom_openapi
```

This code modifies the openapi schema, embeds nested schemas and then removes unused schemas.

Results of embedding `_X` and `_Y` into `Z`:

<details>

```json
"Z": {
  "title": "Z",
  "type": "object",
  "properties": {
    "z": {
      "type": "integer",
      "title": "Z"
    },
    "y_model": {
      "title": "_Y",
      "type": "object",
      "properties": {
        "y": {
          "type": "integer",
          "title": "Y"
        },
        "x_model": {
          "title": "_X",
          "type": "object",
          "properties": {
            "x": {
              "type": "integer",
              "title": "X"
            }
          },
          "required": [
            "x"
          ]
        }
      },
      "required": [
        "y",
        "x_model"
      ]
    }
  },
  "required": [
    "z",
    "y_model"
  ]
}
```

</details>
